File: Medium/Fruit Into Baskets/fruit-into-baskets.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    int totalFruits(int N, vector<int> &fruits) {
        int l=0,r=0,k=2;
        int max_len=0,len=0;
        unordered_map<int,int> mp;
        while(r<N)
        {
            mp[fruits[r]]++;
            
            if(mp.size()>k)
            {
                mp[fruits[l]]--;
                
                if(mp[fruits[l]]==0)
                mp.erase(fruits[l]);
                
                l++;
               
            }
             else if(mp.size()<=k)
            {
                len = r-l+1;
                max_len = max(max_len,len);
                
            }
            r++;
             
        }
        return max_len;
    }
    
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin >> N;
        vector<int> fruits(N);
        for (int i = 0; i < N; i++) cin >> fruits[i];
        Solution obj;
        cout << obj.totalFruits(N, fruits) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Fruit Into Baskets/fruit-into-baskets.py
#User function Template for python3

class Solution:
    def totalFruits(self,arr):
        # Code here
        myMap = {}
        maxLen = -1
        left,right = 0,0
        
        while  right<len(arr):
            if arr[right] in myMap:
                myMap[arr[right]] += 1
            else:
                myMap[arr[right]] = 1
                
            
            if len(myMap)>2:
                
                myMap[arr[left]] -= 1
                if myMap[arr[left]] == 0:
                    del myMap[arr[left]]
                left += 1
            else:
                maxLen = max(maxLen,right-left+1)
                
            right+=1      
           
        return maxLen

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        # N = int(input())
        arr = list(map(int, input().split()))
        ob = Solution()
        res = ob.totalFruits(arr)
        print(res)

# } Driver Code Ends